pkg/cmd/server/kubernetes/master_config.go

```diff
@@ -207,38 +201,32 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
-	// We need to fix the upstream to stop using the storage version as a preferred api version.
-	storageVersions := map[string]string{}
-
-	enabledKubeVersions := configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupKube)
-	if len(enabledKubeVersions) > 0 {
-		kubeStorageVersion := unversioned.GroupVersion{Group: configapi.APIGroupKube, Version: options.EtcdStorageConfig.KubernetesStorageVersion}
-		databaseStorage, err := NewEtcdStorage(etcdClient, kubeStorageVersion, options.EtcdStorageConfig.KubernetesStoragePrefix)
-		if err != nil {
-			return nil, fmt.Errorf("Error setting up Kubernetes server storage: %v", err)
-		}
-		storageDestinations.AddAPIGroup(configapi.APIGroupKube, databaseStorage)
-		storageVersions[configapi.APIGroupKube] = options.EtcdStorageConfig.KubernetesStorageVersion
+	etcdConfig := storagebackend.Config{
+		Prefix:     genericapiserver.DefaultEtcdPathPrefix,
+		ServerList: options.EtcdClientInfo.URLs,
 	}
+	resourceEncodingConfig := genericapiserver.NewDefaultResourceEncodingConfig()
+	resourceEncodingConfig.SetVersionEncoding(
+		configapi.APIGroupKube,
+		unversioned.GroupVersion{Group: configapi.APIGroupKube, Version: options.EtcdStorageConfig.KubernetesStorageVersion},
+		configapi.SchemeGroupVersion,
+	)
 
 	// enable this if extensions API is enabled (or batch or autoscaling, since they persist to extensions/v1beta1 for now)
 	// TODO: replace this with a loop over configured storage versions
 	extensionsEnabled := len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupExtensions)) > 0
 	batchEnabled := len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupBatch)) > 0
 	autoscalingEnabled := len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupAutoscaling)) > 0
 	if extensionsEnabled || autoscalingEnabled || batchEnabled {
-		// TODO: replace this with a configured storage version for extensions once configuration exposes this
-		extensionsStorageVersion := unversioned.GroupVersion{Group: extensions.GroupName, Version: "v1beta1"}
-		databaseStorage, err := NewEtcdStorage(etcdClient, extensionsStorageVersion, options.EtcdStorageConfig.KubernetesStoragePrefix)
-		if err != nil {
-			return nil, fmt.Errorf("Error setting up Kubernetes extensions server storage: %v", err)
-		}
-		storageDestinations.AddAPIGroup(configapi.APIGroupExtensions, databaseStorage)
-		storageVersions[configapi.APIGroupExtensions] = extensionsStorageVersion.String()
+		resourceEncodingConfig.SetVersionEncoding(
+			configapi.APIGroupExtensions,
+			unversioned.GroupVersion{Group: extensions.GroupName, Version: "v1beta1"},
+			extensions.SchemeGroupVersion,
+		)
 	}
 
+	storageFactory := genericapiserver.NewDefaultStorageFactory(etcdConfig, "", kapi.Codecs, resourceEncodingConfig, genericapiserver.NewResourceConfig())
```

Comment:
`APIResourceConfigSource` doesn't look right.  Shouldn't this take into account which groups are actually enabled?